### PR TITLE
Add a script to list updatable packages

### PR DESCRIPTION
**Pull request by evgeni**

### CODE REVIEW

**Comment by evgeni** [comment]
@ekohl you might say this is inspired ;)

File: scripts/list_updatable_packages [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+def main() -> None:
+    matrix = list(build_matrix(find_packages()))
+
+    if 'GITHUB_ACTION' in os.environ:
+        directories = [entry['directory'] for entry in matrix]
+        print(f'::set-output name=directories::{json.dumps(directories)}')
+        print(f'::set-output name=matrix::{json.dumps(matrix)}')
+    for entry in matrix:
+        print(entry['directory'], entry['new_version'])
```
Comment (ekohl) [inline scripts/list_updatable_packages:63]: This should match what's called. In this case it's `./scripts/update_package.rb --name $package_name --version $version`. So `directory` here doesn't make a lot of sense?

Comment (evgeni) [reply scripts/list_updatable_packages:63]: I didn't look too closely at that part (which is copied from you 1:1), but yeah, probably

Comment (evgeni) [reply scripts/list_updatable_packages:63]: looking closer, the directory names in `plugins/` are *not* **always** package names (yay):
* `ruby-foreman-azure-rm` is a package name
* `smart_proxy_chef` is a gem name

So, really, one would need to call `./scripts/update_package.rb --name $(basename $directory) --version $version`

Comment (ekohl) [reply scripts/list_updatable_packages:63]: Looks like that works

File: scripts/list_updatable_packages [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+SESSION = requests.Session()
+
+
+Package = namedtuple('Package', ['package_name', 'version', 'gem_name', 'directory'])
+
+
+def find_packages() -> Generator[Package, None, None]:
+    for plugin_rb in Path('plugins/').glob('**/*.rb'):
+        gem_name = plugin_rb.read_text().strip().split()[1].strip("',\"")
```
Comment (ekohl) [inline scripts/list_updatable_packages:23]: Technically gem_name isn't really used anymore in the action so I wonder if we should drop this. In #7412 we could make the same argument. But I'm fine either way.

File: scripts/list_updatable_packages [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+from debian.changelog import Changelog
+from debian.debian_support import NativeVersion
+
+import requests
+
+SESSION = requests.Session()
+
+
+Package = namedtuple('Package', ['package_name', 'version', 'gem_name', 'directory'])
```
Comment (ekohl) [inline scripts/list_updatable_packages:18]: Recently I've taken to using [dataclasses](https://docs.python.org/3/library/dataclasses.html) but that requires Python 3.7.

File: scripts/list_updatable_packages [hunk 4]
```diff
@@ -0,0 +1,67 @@
+#!/usr/bin/env python3
+
+import json
+import os
+import sys
+from collections import namedtuple
+from pathlib import Path
+from typing import Generator, Iterable, Mapping, Optional
```
Comment (ekohl) [inline scripts/list_updatable_packages:8]: `Mapping` and `Optional` are unused.

File: scripts/list_updatable_packages [hunk 5]
```diff
@@ -0,0 +1,67 @@
+#!/usr/bin/env python3
+
+import json
+import os
+import sys
```
Comment (ekohl) [inline scripts/list_updatable_packages:5]: Unused import

**Review by ekohl — APPROVED** [review]
Thanks! This does work.

File: scripts/list_updatable_packages (hunk 2 above)
Comment (evgeni) [reply scripts/list_updatable_packages:23]: It certainly is needed to ask `rubygems.org` about the version? :)

File: scripts/list_updatable_packages (hunk 3 above)
Comment (evgeni) [reply scripts/list_updatable_packages:18]: What would be the benefit? (besides type hints)

**Review by ezr-ondrej — APPROVED** [review]
Nice :+1:
Can we get that in so I can update my PR to use both? (basically merge @ekohl's additions :))

File: scripts/list_updatable_packages (hunk 3 above)
Comment (ekohl) [reply scripts/list_updatable_packages:17]: The syntax feels more natural and you can have defaults. For example:
```python
@dataclasses.dataclass
class Package:
    package_name: str
    count: int = 0
    mylist: list[int] = dataclasses.field(default_factory=list)
```

The [field()](https://docs.python.org/3/library/dataclasses.html#dataclasses.field) method can also determine if the field is used in the generated methods like `__repr__` and __eq__`.

I quite like it over `namedtuple`, especially if you want to add some additional methods which is always awkward with `namedtuple`.

File: scripts/list_updatable_packages [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    for plugin_rb in Path('plugins/').glob('**/*.rb'):
+        gem_name = plugin_rb.read_text().strip().split()[1].strip("',\"")
+        package_changelog = next(plugin_rb.parent.glob('**/changelog'))
+        with package_changelog.open() as debchangelog:
+            changelog = Changelog(debchangelog)
+            version = NativeVersion(changelog.version).upstream_version
+            yield Package(str(changelog.package), version, gem_name, plugin_rb.parent.as_posix())
+
+
+def latest_version(spec: Package) -> str:
```
Comment (ekohl) [inline scripts/list_updatable_packages:30]: Is `spec` the right name?

Comment (evgeni) [reply scripts/list_updatable_packages:30]: Copy paste from your script ;)

Comment (ekohl) [reply scripts/list_updatable_packages:30]: Well, RPMs have specs. Do Debian packages have specs? ;)

Comment (evgeni) [reply scripts/list_updatable_packages:30]: No. Honestly, I would have used "package" in both cases. The fact we're pulling data from a spec is IMHO irrelevant ;)

**Comment by evgeni** [comment]
Thanks!

**Pull request merged**